**Methods - network analysis - create network data**

_2017.11.14 - work log - prelim - network analysis_

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Setup" data-toc-modified-id="Setup-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Setup</a></span><ul class="toc-item"><li><span><a href="#Setup---Imports" data-toc-modified-id="Setup---Imports-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Setup - Imports</a></span></li><li><span><a href="#Setup---Initialize-Django" data-toc-modified-id="Setup---Initialize-Django-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Setup - Initialize Django</a></span></li></ul></li><li><span><a href="#Network-Analysis" data-toc-modified-id="Network-Analysis-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Network Analysis</a></span><ul class="toc-item"><li><span><a href="#networks---test-prelim-network---original-coders" data-toc-modified-id="networks---test-prelim-network---original-coders-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>networks - test prelim network - original coders</a></span></li><li><span><a href="#networks---test-prelim-network---new-coders" data-toc-modified-id="networks---test-prelim-network---new-coders-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>networks - test prelim network - new coders</a></span></li><li><span><a href="#networks---test-prelim-month" data-toc-modified-id="networks---test-prelim-month-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>networks - test prelim month</a></span><ul class="toc-item"><li><span><a href="#nodes---full-month;-ties---full-month" data-toc-modified-id="nodes---full-month;-ties---full-month-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>nodes - full month; ties - full month</a></span></li><li><span><a href="#nodes---full-month;-ties---week-of-2009-12-06-to-2009-12-12" data-toc-modified-id="nodes---full-month;-ties---week-of-2009-12-06-to-2009-12-12-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>nodes - full month; ties - week of 2009-12-06 to 2009-12-12</a></span></li><li><span><a href="#TODO---nodes---full-month;-ties---week-of-2009-12-13-to-2009-12-19" data-toc-modified-id="TODO---nodes---full-month;-ties---week-of-2009-12-13-to-2009-12-19-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>TODO - nodes - full month; ties - week of 2009-12-13 to 2009-12-19</a></span></li><li><span><a href="#TODO---nodes---full-month;-ties---week-of-2009-12-20-to-2009-12-26" data-toc-modified-id="TODO---nodes---full-month;-ties---week-of-2009-12-20-to-2009-12-26-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>TODO - nodes - full month; ties - week of 2009-12-20 to 2009-12-26</a></span></li></ul></li></ul></li><li><span><a href="#Redo-network-analysis" data-toc-modified-id="Redo-network-analysis-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Redo network analysis</a></span><ul class="toc-item"><li><span><a href="#analysis-person_info.py" data-toc-modified-id="analysis-person_info.py-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>analysis-person_info.py</a></span></li><li><span><a href="#R-network-analysis" data-toc-modified-id="R-network-analysis-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>R network analysis</a></span></li></ul></li><li><span><a href="#TODO" data-toc-modified-id="TODO-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>TODO</a></span></li></ul></div>

# Setup

- Back to [Table of Contents](#Table-of-Contents)

## Setup - Imports

- Back to [Table of Contents](#Table-of-Contents)

In [1]:
from __future__ import unicode_literals
from __future__ import division

# python base imports
import datetime

# import six
import six

print( "packages imported at " + str( datetime.datetime.now() ) )

packages imported at 2018-03-17 21:55:28.967228


## Setup - Initialize Django

- Back to [Table of Contents](#Table-of-Contents)

First, initialize my dev django project, so I can run code in this notebook that references my django models and can talk to the database using my project's settings.

In [2]:
%run ../django_init.py

/home/jonathanmorgan/.virtualenvs/sourcenet/lib/python3.5/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


django initialized at 2018-03-18 01:55:33.353472


In [3]:
# django imports
from django.contrib.auth.models import User

# sourcenet imports
from sourcenet.shared.sourcenet_base import SourcenetBase

# sourcenet_analysis imports
from sourcenet_analysis.network.network_person_info import NetworkPersonInfo

# sourcenet imports
from sourcenet.models import Article
from sourcenet.models import Article_Data
from sourcenet.models import Person


# Network Analysis

- Back to [Table of Contents](#Table-of-Contents)

Generate some basic network statistics from the ground truth and automated attribution data, characterize and compare using QAP (including explaining substantial limitations of this given sparseness of networks).

- find code used to derive network information last time. - Evernote Network Analysis Notes: [https://www.evernote.com/shard/s101/nl/11379781/ef9db83f-5fd3-4bd2-bdd7-1407e2c01f9c/](https://www.evernote.com/shard/s101/nl/11379781/ef9db83f-5fd3-4bd2-bdd7-1407e2c01f9c/)
- run it again on full month of data, rather than just a week.
- examine traits of ground_truth and automated networks

    - compare with QAP.
    - look at average degree of reporters
    - look at average degree, density, etc.

Notes:

- TK

## networks - test prelim network - original coders

- Back to [Table of Contents](#Table-of-Contents)

First, we configure as we did before, render human and automated networks, and compare the files to the previous run - should be identical.

Configuration:

* Configuration to generate network files for prelim:

    * Config of "Select Articles" - use defaults, except:

        * People from: 2009-12-06
        * People to: 2009-12-12
        * Publications: "Grand Rapids Press, The"
        * Coders:

            * for human sample: "brianbowe", "jonathanmorgan"
            * for automated: "automated"

        * `coder_type filter` - only for coder "automated", for now.

            * use the coder_type filter fields to filter automatically coded Article_Data on coder type if you have tried different automated coder types:
            * "Article_Data coder_type Filter Type:" - "Just automated"
            * "coder_type 'Value In' List (comma-delimited):" - Enter the coder types you want included.  Examples:
                
                * OpenCalais v2: "OpenCalais_REST_API_v2"
        
        * Article Tag List: "prelim_network"
        * Person allow duplicate articles: "No"
    
    * Configure "Network Settings" - use defaults, except:
        
        * Download as File?: "Yes"
        * Data Format: "Tab-Delimited Matrix"
        * Data Output Type: Network + Attribute Columns
        * Include Headers: "Yes"
    
    * Config of "Select People" - use defaults, except:
        
        * For each, include all people used by either human or automated, so you have same dimensions.  SO...
        * Person Query Type: "Custom, defined below"
        * People from: 2009-12-06
        * People to: 2009-12-12
        * Person publications: "Grand Rapids Press, The"
        * Person coders: "automated", "brianbowe", "jonathanmorgan"
        * coder_type filter
            
            * use the coder_type filter fields to filter automatically coded Article_Data on coder type if you have tried different automated coder types:
            * "Article_Data coder_type Filter Type:" - "Just automated"
            * "coder_type 'Value In' List (comma-delimited):" - Enter the coder types you want included.  Examples:
                
                * OpenCalais v2: "OpenCalais_REST_API_v2"
        
        * Article Tag List: "prelim_network"
        * Person allow duplicate articles: "Yes"


Resulting files stored in Dropbox (`Dropbox/academia/MSU/program_stuff/prelim_paper/data/network_analysis/2017.11.14/network/original_coders/prelim_network`):

- human - original coders - article and person unordered - `sourcenet_data-20171114-182817-prelim_network-week-human-original-unordered.tab`
- human - original coders - article ordered (6,4), person unordered - `sourcenet_data-20171114-183930-prelim_network-week-human-original-ordered-64.tab`
- human - original coders - article (6,4) and person (2,6,4) ordered - `sourcenet_data-20171115-024247-prelim_network-week-human-original-ordered-64-264.tab`
- automated - `sourcenet_data-20171114-182942-prelim_network-week-automated.tab`

Notes:

- The human ones are almost identical - compared in kaleidoscope - 3 lines different.  The automated one is quite different - evidence that network correlations are not really so useful with so many 0s.
- Ordered or unordered in person, configured as it was for this, files are identical (YAY! as it should be!).

## networks - test prelim network - new coders

- Back to [Table of Contents](#Table-of-Contents)

Next, we start with orignal configuration, update to reflect new coder users and ground_truth, and render human and automated networks for the same week, and compare the files to the old coders.

Configuration to test with new coders:

- start with base configuration above.
- set up the Person so it is using minnesota1, minnesota2, minnesota3, ground_truth, and automated, unordered.
- in Article:

    - try automated alone.
    - try minnesota1, minnesota2, minnesota3, ground_truth unordered.
    - try ordered, ground_truth, then minnesota1, minnesota2, minnesota3 - 13, 8, 9, 10

Resulting files stored in Dropbox (`Dropbox/academia/MSU/program_stuff/prelim_paper/data/network_analysis/2017.11.14/network/new_coders/prelim_network`):

- automated - `sourcenet_data-20171115-033413-new-prelim_network-week-automated.tab`
- automated using ordered fields - `sourcenet_data-20171115-041030-new-prelim_network-week-automated-all_ordered-person-2.13.8.9.10.tab`
- human - new coders - unordered - `sourcenet_data-20171115-035924-new-prelim_network-week-human-unordered.tab`
- human - new coders - ordered 13,8,9,10 - `sourcenet_data-20171114-183930-new-prelim_network-week-human-new-ordered-13.8.9.10.tab`

Notes:

- There are differences, but they are about what I'd expect for ground_truth taking precedence consistently - 35 rows different.
- automated I tried two ways - using select, and using ordered list.  Identical either way.

## networks - test prelim month

- Back to [Table of Contents](#Table-of-Contents)

Next, we look at how prelim_month automated and user compare.


### nodes - full month; ties - full month

Configuration:

* Configuration to generate network files for prelim:

    * Config of "Select Articles" - use defaults, except:

        * People from: 2009-12-01
        * People to: 2009-12-31
        * Publications: "Grand Rapids Press, The"
        * Coders: empty
        * Coder IDs to include, in order of highest to lowest priority: 

            * for human sample: 13,8,9,10
            * for automated: 2

        * `coder_type filter` - only for coder "automated", for now.

            * use the coder_type filter fields to filter automatically coded Article_Data on coder type if you have tried different automated coder types:
            * "Article_Data coder_type Filter Type:" - "Just automated"
            * "coder_type 'Value In' List (comma-delimited):" - Enter the coder types you want included.  Examples:
                
                * OpenCalais v2: "OpenCalais_REST_API_v2"
        
        * Article Tag List: "grp_month"
        * Person allow duplicate articles: "No"
    
    * Configure "Network Settings" - use defaults, except:
        
        * Download as File?: "Yes"
        * Data Format: "Tab-Delimited Matrix"
        * Data Output Type: Network + Attribute Columns
        * Include Headers: "Yes"
    
    * Config of "Select People" - use defaults, except:
        
        * For each, include all people used by either human or automated, so you have same dimensions.  SO...
        * Person Query Type: "Custom, defined below"
        * People from: 2009-12-01
        * People to: 2009-12-31
        * Person publications: "Grand Rapids Press, The"
        * Person coders: "automated", "minnesota1", "minnesota2", "minnesota3", "ground_truth"
        * coder_type filter
            
            * use the coder_type filter fields to filter automatically coded Article_Data on coder type if you have tried different automated coder types:
            * "Article_Data coder_type Filter Type:" - "Just automated"
            * "coder_type 'Value In' List (comma-delimited):" - Enter the coder types you want included.  Examples:
                
                * OpenCalais v2: "OpenCalais_REST_API_v2"
        
        * Article Tag List: "grp_month"
        * Person allow duplicate articles: "Yes"

Resulting files stored in Dropbox (`Dropbox/academia/MSU/program_stuff/prelim_paper/data/network_analysis/2017.11.14/network/new_coders/grp_month`):

- automated - `sourcenet_data-20171205-022551-grp_month-automated.tab`
- human - `sourcenet_data-20171115-043102-grp_month-human.tab`


### nodes - full month; ties - week of 2009-12-06 to 2009-12-12

Then, alter to just output the week, but with all the people for the entire month (so looking at how same matrix compares when populated using a week's worth of data compared to a month's).

* Configuration to generate network files for prelim - use the above grp_month config, except:

    * Config of "Select Articles" - use defaults, except:

        * People from: 2009-12-06
        * People to: 2009-12-12

Resulting files stored in Dropbox (`Dropbox/academia/MSU/program_stuff/prelim_paper/data/network_analysis/2017.11.14/network/prelim_network/new_coders/grp_month`):

- automated - `sourcenet_data-20171206-031358-grp_month-automated-week_subset.tab`
- human - `sourcenet_data-20171206-031319-grp_month-human-week_subset.tab`


### TODO - nodes - full month; ties - week of 2009-12-13 to 2009-12-19

Then, alter to just output the week, but with all the people for the entire month (so looking at how same matrix compares when populated using a week's worth of data compared to a month's).

* Configuration to generate network files for prelim - use the above grp_month config, except:

    * Config of "Select Articles" - use defaults, except:

        * People from: 2009-12-13
        * People to: 2009-12-19

Resulting files stored in Dropbox (`Dropbox/academia/MSU/program_stuff/prelim_paper/data/network_analysis/2017.11.14/network/prelim_network/new_coders/grp_month`):

- automated - `sourcenet_data-20171206-031358-grp_month-automated-week2_subset.tab`
- human - `sourcenet_data-20171206-031319-grp_month-human-week2_subset.tab`


### TODO - nodes - full month; ties - week of 2009-12-20 to 2009-12-26

Then, alter to just output the week, but with all the people for the entire month (so looking at how same matrix compares when populated using a week's worth of data compared to a month's).

* Configuration to generate network files for prelim - use the above grp_month config, except:

    * Config of "Select Articles" - use defaults, except:

        * People from: 2009-12-20
        * People to: 2009-12-26

Resulting files stored in Dropbox (`Dropbox/academia/MSU/program_stuff/prelim_paper/data/network_analysis/2017.11.14/network/prelim_network/new_coders/grp_month`):

- automated - `sourcenet_data-20171206-031358-grp_month-automated-week3_subset.tab`
- human - `sourcenet_data-20171206-031319-grp_month-human-week3_subset.tab`


# Redo network analysis

- Back to [Table of Contents](#Table-of-Contents)

Notes on original network analysis are in Evernote: [MSU PhD - prelim - analysis - Network Analysis notes](https://www.evernote.com/l/AGXvnbg_X9NL0r3XFAfiwB-cgnAnalb1vwg)

* network descriptives
    * high-level description of networks
    * // From Prelim Proposal: "I’ll then compare values for more specific network statistics between the two, including network-level aggregate degree and transitivity, and reporter-level degree, average tie weight and transitivity."  More details:
    * http://faculty.ucr.edu/~hanneman/nettext/C7_Connection.html
    * for old results and more detailed notes on implementation and interpretation, see `Dropbox/academia/MSU/program_stuff/prelim_paper/analysis/analysis_summary.xlsx`
    * network-level
        
        * files
            
            * python script:
                
                * `sourcenet/examples/analysis/analysis-person_info.py` - calculates per-author information - on shared sources, article counts per author, etc.
            
            * R scripts:
                
                * `sourcenet/R/db_connect.r`
                * `sourcenet/R/sna/functions-sna.r`
                * `sourcenet/R/sna/sna-load_data.r`
                * `sourcenet/R/sna/igraph/*`
                * `sourcenet/R/sna/statnet/*`
        
        * statnet/sna
            
            * `sna::gden()` - graph density
        
        * igraph
            
            * `igraph::transitivity()` - vector of transitivity scores for each node in a graph, plus network-level transitivity score.
                
                * Q - interpretation?

## analysis-person_info.py

- Back to [Table of Contents](#Table-of-Contents)

First, need to figure out `sourcenet/examples/analysis/reliability-build_relations.py`

Original file: `sourcenet/examples/analysis/analysis-person_info.py`

Moved to:

- NetworkPersonInfo object: `sourcenet_analysis/network/network_person_info.py`
- example file: `sourcenet_analysis/examples/network/network-person_info.py`

Try reproducing below with new class - Configure:

In [4]:
%run ../config-coder_index-prelim_month.py

indexing for grp_month/prelim_month initialized at 2018-03-18 01:58:02.409032


In [5]:
%run ../config-coder_index-prelim_week.py

indexing for prelim_network (1 week) initialized at 2018-03-18 01:58:04.341522


And then run the code:

In [7]:
#===============================================================================
# process articles
#===============================================================================

# process articles
my_info_instance.process_articles( tag_list )

#output lists of counts of sources and shared source by author

# declare variables - looking at data
coder_index_to_data_dict = None
coder_index = -1
coder_data_dict = None
coder_author_id_list = None
coder_author_source_count_list = None
coder_author_shared_count_list = None
coder_author_article_count_list = None
mean_source_count = -1
mean_shared_count = -1
mean_article_count = -1
author_index = -1
shared_count = -1
temp_author_id_list = []
temp_source_count_list = []
temp_shared_count_list = []
temp_article_count_list = []

# for each index, get authors.
coder_index_to_data_dict = my_info_instance.coder_index_to_data_map
        
# loop over the dictionary to process each index.
for coder_index, coder_data_dict in six.iteritems( coder_index_to_data_dict ):

    # get data for coder
    coder_author_id_list = coder_data_dict.get( NetworkPersonInfo.PROP_CODER_AUTHOR_ID_LIST, None )
    coder_author_source_count_list = coder_data_dict.get( NetworkPersonInfo.PROP_CODER_AUTHOR_SOURCE_COUNT_LIST, None )
    coder_author_shared_count_list = coder_data_dict.get( NetworkPersonInfo.PROP_CODER_AUTHOR_SHARED_COUNT_LIST, None )
    coder_author_article_count_list = coder_data_dict.get( NetworkPersonInfo.PROP_CODER_AUTHOR_ARTICLE_COUNT_LIST, None )

    # output
    print( "" )
    print( "================================================================================" )
    print( "Data for Coder index " + str( coder_index ) + ":" )

    print( "" )
    print( "==> All authors" )
    print( "- author ID list = " + str( coder_author_id_list ) )    
    print( "- author source count list = " + str( coder_author_source_count_list ) )    
    print( "- author shared count list = " + str( coder_author_shared_count_list ) )    
    print( "- author article count list = " + str( coder_author_article_count_list ) )    

    # and some computations

    # author count
    print( "- author count = " + str( len( coder_author_id_list ) ) )
    
    # mean source count per author
    mean_source_count = float( sum( coder_author_source_count_list ) ) / len( coder_author_source_count_list )
    print( "- mean source count per author = " + str( mean_source_count ) )
    
    # mean shared count per author
    mean_shared_count = float( sum( coder_author_shared_count_list ) ) / len( coder_author_shared_count_list )
    print( "- mean shared count per author = " + str( mean_shared_count ) )
    
    # mean article count per author
    mean_article_count = float( sum( coder_author_article_count_list ) ) / len( coder_author_article_count_list )
    print( "- mean article count per author = " + str( mean_article_count ) )
    
    # the same, but just for those with shared sources.
    author_index = -1
    temp_author_id_list = []
    temp_source_count_list = []
    temp_shared_count_list = []
    temp_article_count_list = []

    for shared_count in coder_author_shared_count_list:
    
        # increment index
        author_index += 1
        
        # greater than 0?
        if ( shared_count > 0 ):
        
            # yes, add info to temp lists.
            temp_author_id_list.append( coder_author_id_list[ author_index ] )
            temp_source_count_list.append( coder_author_source_count_list[ author_index ] )
            temp_shared_count_list.append( coder_author_shared_count_list[ author_index ] )
            temp_article_count_list.append( coder_author_article_count_list[ author_index ] )
            
        #-- END check to see if shared count > 0 --#
    
    #-- END loop over shared_count_list --#

    print( "" )
    print( "==> Authors with shared sources" )
    print( "- author ID list = " + str( temp_author_id_list ) )    
    print( "- author source count list = " + str( temp_source_count_list ) )    
    print( "- author shared count list = " + str( temp_shared_count_list ) )    
    print( "- author article count list = " + str( temp_article_count_list ) )    

    # and some computations

    # author count
    print( "- author count = " + str( len( temp_author_id_list ) ) )
    
    # mean source count per author with shared sources
    mean_source_count = float( sum( temp_source_count_list ) ) / len( temp_source_count_list )
    print( "- mean source count per author with shared sources = " + str( mean_source_count ) )
    
    # mean shared count per author with shared sources
    mean_shared_count = float( sum( temp_shared_count_list ) ) / len( temp_shared_count_list )
    print( "- mean shared count per author with shared sources = " + str( mean_shared_count ) )
    
    # mean article count per author
    mean_article_count = float( sum( temp_article_count_list ) ) / len( temp_article_count_list )
    print( "- mean article count per author = " + str( mean_article_count ) )
    
#-- END loop over coders. --#


Start of update_author_shared_sources():
In update_author_shared_sources: multiple authors ( [3, 84] for source 2731
In update_author_shared_sources: multiple authors ( [598, 1082] for source 2052
In update_author_shared_sources: multiple authors ( [598, 1082] for source 2054
In update_author_shared_sources: multiple authors ( [598, 1082] for source 2055
In update_author_shared_sources: multiple authors ( [598, 1082] for source 2056
In update_author_shared_sources: multiple authors ( [69, 84] for source 345
In update_author_shared_sources: multiple authors ( [377, 69, 23, 437] for source 2094
In update_author_shared_sources: multiple authors ( [66, 23] for source 2096
In update_author_shared_sources: multiple authors ( [598, 1082] for source 2060
In update_author_shared_sources: multiple authors ( [377, 69] for source 100
In update_author_shared_sources: multiple authors ( [46, 161] for source 358
In update_author_shared_sources: multiple authors ( [66, 223, 46, 69, 36, 23] for source

Results for:

- grp_month/prelim_month: msu_phd_work/results/network_person_info-grp_month.txt

        Processed 441 Articles.
        Processed 882 Article_Data records.

        ================================================================================
        Data for Coder index 1:

        ==> All authors
        - author ID list = [387, 2310, 2567, 394, 652, 13, 654, 3, 46, 23, 2004, 29, 30, 417, 36, 425, 2614, 302, 178, 437, 566, 1082, 443, 377, 66, 69, 161, 73, 74, 460, 482, 591, 336, 84, 598, 599, 217, 223, 736, 2018, 743, 937, 1782, 1655, 332, 505, 703, 637]
        - author source count list = [18, 2, 0, 33, 9, 36, 3, 27, 57, 31, 4, 50, 28, 4, 31, 30, 5, 31, 41, 45, 4, 13, 43, 36, 92, 43, 37, 30, 46, 3, 1, 76, 9, 64, 21, 50, 46, 18, 2, 5, 2, 7, 4, 6, 7, 18, 2, 13]
        - author shared count list = [7, 2, 0, 2, 0, 1, 0, 9, 22, 12, 0, 2, 2, 0, 9, 2, 0, 3, 6, 13, 0, 5, 9, 10, 37, 19, 12, 10, 5, 1, 0, 6, 2, 19, 5, 4, 13, 9, 0, 0, 0, 7, 0, 6, 1, 1, 0, 0]
        - author article count list = [7, 1, 1, 8, 5, 17, 1, 13, 21, 15, 2, 18, 13, 4, 11, 10, 1, 12, 13, 15, 1, 8, 16, 17, 30, 15, 14, 12, 19, 4, 1, 25, 4, 27, 9, 17, 18, 6, 1, 1, 1, 1, 4, 1, 4, 8, 2, 4]
        - author count = 48
        - mean source count per author = 24.645833333333332
        - mean shared count per author = 5.6875
        - mean article count per author = 9.541666666666666

        ==> Authors with shared sources
        - author ID list = [387, 2310, 394, 13, 3, 46, 23, 29, 30, 36, 425, 302, 178, 437, 1082, 443, 377, 66, 69, 161, 73, 74, 460, 591, 336, 84, 598, 599, 217, 223, 937, 1655, 332, 505]
        - author source count list = [18, 2, 33, 36, 27, 57, 31, 50, 28, 31, 30, 31, 41, 45, 13, 43, 36, 92, 43, 37, 30, 46, 3, 76, 9, 64, 21, 50, 46, 18, 7, 6, 7, 18]
        - author shared count list = [7, 2, 2, 1, 9, 22, 12, 2, 2, 9, 2, 3, 6, 13, 5, 9, 10, 37, 19, 12, 10, 5, 1, 6, 2, 19, 5, 4, 13, 9, 7, 6, 1, 1]
        - author article count list = [7, 1, 8, 17, 13, 21, 15, 18, 13, 11, 10, 12, 13, 15, 8, 16, 17, 30, 15, 14, 12, 19, 4, 25, 4, 27, 9, 17, 18, 6, 1, 1, 4, 8]
        - author count = 34
        - mean source count per author with shared sources = 33.088235294117645
        - mean shared count per author with shared sources = 8.029411764705882
        - mean article count per author = 12.617647058823529

        ================================================================================
        Data for Coder index 2:

        ==> All authors
        - author ID list = [387, 2310, 2567, 394, 652, 13, 654, 3, 46, 23, 2004, 29, 30, 417, 36, 425, 2614, 302, 178, 437, 566, 1082, 443, 377, 66, 69, 161, 73, 74, 460, 482, 591, 336, 84, 598, 599, 217, 223, 736, 2018, 743, 1782, 1655, 332, 505, 703, 637]
        - author source count list = [18, 2, 0, 27, 8, 39, 2, 29, 46, 33, 4, 50, 26, 4, 28, 31, 6, 31, 42, 49, 2, 15, 43, 34, 88, 45, 34, 28, 46, 4, 1, 72, 9, 69, 22, 46, 43, 13, 2, 5, 2, 4, 6, 7, 14, 2, 10]
        - author shared count list = [7, 2, 0, 2, 0, 1, 0, 12, 13, 11, 0, 0, 2, 0, 7, 3, 1, 4, 8, 10, 0, 7, 8, 9, 35, 19, 11, 10, 4, 1, 0, 6, 1, 20, 7, 3, 11, 8, 0, 0, 0, 0, 6, 1, 1, 0, 0]
        - author article count list = [7, 1, 1, 8, 5, 17, 1, 13, 20, 15, 2, 18, 13, 4, 11, 10, 1, 12, 13, 15, 1, 8, 16, 17, 30, 15, 14, 12, 19, 4, 1, 25, 4, 27, 9, 17, 18, 6, 1, 1, 1, 4, 1, 4, 8, 2, 4]
        - author count = 47
        - mean source count per author = 24.27659574468085
        - mean shared count per author = 5.340425531914893
        - mean article count per author = 9.702127659574469

        ==> Authors with shared sources
        - author ID list = [387, 2310, 394, 13, 3, 46, 23, 30, 36, 425, 2614, 302, 178, 437, 1082, 443, 377, 66, 69, 161, 73, 74, 460, 591, 336, 84, 598, 599, 217, 223, 1655, 332, 505]
        - author source count list = [18, 2, 27, 39, 29, 46, 33, 26, 28, 31, 6, 31, 42, 49, 15, 43, 34, 88, 45, 34, 28, 46, 4, 72, 9, 69, 22, 46, 43, 13, 6, 7, 14]
        - author shared count list = [7, 2, 2, 1, 12, 13, 11, 2, 7, 3, 1, 4, 8, 10, 7, 8, 9, 35, 19, 11, 10, 4, 1, 6, 1, 20, 7, 3, 11, 8, 6, 1, 1]
        - author article count list = [7, 1, 8, 17, 13, 20, 15, 13, 11, 10, 1, 12, 13, 15, 8, 16, 17, 30, 15, 14, 12, 19, 4, 25, 4, 27, 9, 17, 18, 6, 1, 4, 8]
        - author count = 33
        - mean source count per author with shared sources = 31.666666666666668
        - mean shared count per author with shared sources = 7.606060606060606
        - mean article count per author = 12.424242424242424

- prelim_network (1 week): msu_phd_work/results/network_person_info-prelim_network-1week.txt

        Processed 109 Articles.
        Processed 214 Article_Data records.

        ================================================================================
        Data for Coder index 1:

        ==> All authors
        - author ID list = [66, 387, 69, 73, 74, 23, 332, 13, 591, 336, 3, 84, 302, 599, 217, 29, 30, 223, 161, 36, 425, 46, 178, 937, 505, 443, 460, 394, 377]
        - author source count list = [17, 5, 23, 6, 16, 10, 3, 14, 23, 9, 6, 19, 15, 8, 11, 11, 10, 7, 16, 17, 5, 32, 13, 7, 4, 4, 0, 7, 2]
        - author shared count list = [12, 0, 6, 1, 1, 7, 0, 0, 1, 0, 0, 7, 2, 0, 1, 1, 0, 7, 9, 7, 2, 16, 0, 7, 0, 0, 0, 0, 0]
        - author article count list = [4, 2, 9, 2, 8, 3, 1, 5, 6, 3, 4, 6, 6, 4, 6, 4, 3, 1, 5, 5, 2, 11, 6, 1, 1, 1, 1, 1, 2]
        - author count = 29
        - mean source count per author = 11.03448275862069
        - mean shared count per author = 3.0
        - mean article count per author = 3.896551724137931

        ==> Authors with shared sources
        - author ID list = [66, 69, 73, 74, 23, 591, 84, 302, 217, 29, 223, 161, 36, 425, 46, 937]
        - author source count list = [17, 23, 6, 16, 10, 23, 19, 15, 11, 11, 7, 16, 17, 5, 32, 7]
        - author shared count list = [12, 6, 1, 1, 7, 1, 7, 2, 1, 1, 7, 9, 7, 2, 16, 7]
        - author article count list = [4, 9, 2, 8, 3, 6, 6, 6, 6, 4, 1, 5, 5, 2, 11, 1]
        - author count = 16
        - mean source count per author with shared sources = 14.6875
        - mean shared count per author with shared sources = 5.4375
        - mean article count per author = 4.9375

        ================================================================================
        Data for Coder index 2:

        ==> All authors
        - author ID list = [66, 387, 69, 73, 74, 23, 332, 13, 591, 336, 505, 3, 340, 46, 599, 217, 29, 30, 223, 84, 161, 36, 425, 566, 302, 178, 350, 758, 377, 443, 460, 394]
        - author source count list = [15, 5, 23, 6, 15, 10, 3, 14, 21, 9, 3, 5, 2, 23, 8, 10, 10, 9, 5, 17, 14, 14, 5, 1, 16, 12, 1, 1, 2, 4, 1, 6]
        - author shared count list = [10, 0, 5, 0, 0, 6, 0, 0, 1, 0, 0, 0, 0, 7, 0, 0, 0, 0, 5, 6, 7, 5, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0]
        - author article count list = [4, 2, 9, 2, 8, 3, 1, 5, 6, 3, 1, 4, 1, 10, 4, 6, 4, 3, 1, 6, 5, 5, 2, 1, 6, 6, 1, 1, 2, 1, 1, 1]
        - author count = 32
        - mean source count per author = 9.0625
        - mean shared count per author = 1.75
        - mean article count per author = 3.59375

        ==> Authors with shared sources
        - author ID list = [66, 69, 23, 591, 46, 223, 84, 161, 36, 425, 302]
        - author source count list = [15, 23, 10, 21, 23, 5, 17, 14, 14, 5, 16]
        - author shared count list = [10, 5, 6, 1, 7, 5, 6, 7, 5, 2, 2]
        - author article count list = [4, 9, 3, 6, 10, 1, 6, 5, 5, 2, 6]
        - author count = 11
        - mean source count per author with shared sources = 14.818181818181818
        - mean shared count per author with shared sources = 5.090909090909091
        - mean article count per author = 5.181818181818182

## R network analysis

- Back to [Table of Contents](#Table-of-Contents)

Notebooks:

- [2017.11.14-work_log-prelim-network_analysis.ipynb](2017.11.14-work_log-prelim-network_analysis.ipynb) - original Python network analysis, run for original coders, then week and month of new data - per-author source, shared, and article counts; and means of each across all authors.
- [2017.12.02-work_log-prelim-R-igraph-grp_month.ipynb](2017.12.02-work_log-prelim-R-igraph-grp_month.ipynb) - basic network analysis of new month and week using igraph.
- [2017.12.02-work_log-prelim-R-statnet-grp_month.ipynb](2017.12.02-work_log-prelim-R-statnet-grp_month.ipynb) - basic network analysis of new month and week using statnet.
- [2017.12.07-work_log-prelim-R-grp_month-sna-author_info.r.ipynb](2017.12.07-work_log-prelim-R-grp_month-sna-author_info.r.ipynb) - R-based author info (similar to Python-based above) for full month of data.
- [2017.12.07-work_log-prelim-R-statnet-grp_month-01.ipynb](2017.12.07-work_log-prelim-R-statnet-grp_month-01.ipynb) - statnet analysis when network is converted to either be 0 or 1 weight, where all weights greater than 1 are converted to 1.  No real difference here, so ignoring in paper.
- [2017.12.09-work_log-prelim-R-grp_week-sna-author_info.r.ipynb](2017.12.09-work_log-prelim-R-grp_week-sna-author_info.r.ipynb) - R-based author info (similar to Python-based above) for single week of data.

# TODO

- Back to [Table of Contents](#Table-of-Contents)

TODO:

- TK

DONE:

- updated forms `ArticleSelectForm` and `PersonSelectForm` to include field for "`coder_id_priority_list`"/"`person_coder_id_priority_list`".
- created method NetworkOutput.get_coder_id_list() that:

    - knows about the two places where coder IDs can be set.
    - if prioritzed list is present:
    
        - starts with the priotized list
        - appends coders from other field who aren't already in the list to the end of it.
        - stores the list in an instance variable inside the object so it can be retrieved easily.
        
- updated NetworkOutput.create_query_set() to use get_coder_id_list() method.
- need to update NetworkOutput.remove_duplicate_article_data() - it is where we choose which Article_Data to omit per article where there are duplicates.  Need to go with order of list.  Might already do this...  Nope.

    - get prioritized list.
    - for first instance of Article_Data for article, store it (related by id, or unique_identifier?)
    - on subsequent Article_Data for article, get index of coder for existing and new.
    - Whichever has lower index you keep.
    - **_Need to test_**

        - person-coded articles:
        
            - 1) output networks from initial prelim (12/6/2009-12/13/2009).  Make sure they are the same now as they were then.
            - 2) keep article specs the same, but change person lookup to use ordered ID list.  See if this is the same as the files in 1 (it is!).  NO NEED: If not, count rows, find and compare rows for some users to see how different they are.  Hopefully same contents, different order...?
            - 3) then, switch the article specs to use ordered list and put person specs back to old way, see how this file compares to the others.  3 lines different.
            - look for differences in:
            
                - number of rows
                - contents of rows
                - IDs of those included
            
        - automated coder:

            - 1) output networks from initial prelim (12/6/2009-12/13/2009).  Make sure they are the same now as they were then.
            - 2) keep article specs the same, but change person lookup to use ordered ID list.  See if this is the same as the files in 1 (might not be).  If not, count rows, find and compare rows for some users to see how different they are.  Hopefully same contents, different order.
            - 3) then, switch the article specs to use ordered list, see how this file compares to the others.

        - as long as the tests above check out, then try out the whole month, with prioritized coder list.

- need to update NetworkDataOutput and children?  Looks like no - all comes down to the remove_duplicate_article_data().
- make sure the network code can deal with multiple coders, and can prioritize in an order I specify.
- figure out the old network analysis stuff.

    - get `sourcenet/examples/analysis/analysis-person_info.py` moved into sourcenet_analysis and working with new index ordering code (might be enough to just extend Reliability_Names_Builder for init and for index specing, then override `process_articles()`.